### ISSUE TITLE
rehydrateSheet - very long

### ISSUE THREAD

**Issue by RDemian** [styled-components/styled-components#3197]
Hi, we use styled-components with SSR and Next.js When loading our site, on the client's side, a significant amount of time is taken by rehydrateSheet. At the same time, we have set up an SSR for styled-components and styles come from the server and are displayed by the browser before scripts are executed on the client side. Is it possible to reduce the hydration time of styles on the client side? How to reduce the number of insertBefore?

our site: [https://perm.domru.ru/](url)

![rehydrateSheet](https://user-images.githubusercontent.com/33232492/87011854-e579f980-c1e1-11ea-8c2b-56ca4863f023.png)

This example also has insertBefore operations for building CSSOM: https://github.com/vercel/next.js/tree/master/examples/with-styled-components

**Comment by probablyup** [comment]
Is your source code accessible somewhere so I can profile?

**Comment by RDemian** [comment]
Sorry, the source code is not available

**Comment by jancajak** [comment]
Hello, we experience same problem. 
Our codebase is not available. It is SSR e-commerce platform and we render page with many elements on server. Server generates style tag with special attributes for styled-components, then rehydrateSheet is looping through this tag content.
![Snímek obrazovky 2020-09-15 v 14 53 02](https://user-images.githubusercontent.com/37411618/93212653-2a328b80-f763-11ea-981b-ac79bb0b9398.png)
We are trying to optimize our site for slower mobile devices. Screenshot is on 4x times slowed CPU.
I am interested if there is any way we can reduce this main thread time. 

Thanks for answer
Jakub

**Comment by probablyup** [comment]
It's impossible to help you without some sort of reproduction, sorry.

**Comment by jancajak** [comment]
Ok, I will try to create minimal reproduction repo here on github.

**Comment by vitalyzhakov** [comment]
@probablyup , @RDemian says that

> This example also has insertBefore operations for building CSSOM:
> https://github.com/vercel/next.js/tree/master/examples/with-styled-components

**Comment by probablyup** [comment]
We only use `insertBefore` for TextTag (development) and creating a single stylesheet during rehydration. I can't see how that would make things extremely slow since it's a single (quite lightweight) DOM operation.

**Comment by jancajak** [comment]
I came into this conclusion by profiler. It shows quite a long times spent by this operation. If I do production build and omit this operation TBT is improved by 100ms. I Expect problem is in our source code somewhere, as reproducible repo do not show any signs of same problem. Do you have any clues what can cause such long duration of normally lightweigth operation?

**Comment by jancajak** [comment]
I created a small sample

`const StyledDiv = styled.div<{ background: string }>`
  background-color: #${(props) => props.background}
`;

const StyledDiv = styled.div.attrs((props) => ({
  style: {
    backgroundColor: `#${props.background}`,
  },
}))``;

export default () => {
  return (
    <div>
      <GlobalStyle />
      <ThemeProvider theme={theme}>
        {Array(1000).fill('').map((_, i) => (
          <StyledDiv key={i} background={Math.floor(Math.random()*16777215).toString(16)}>Hello</StyledDiv>
        ))}
      </ThemeProvider>
    </div>
  );
};`

If I use first version of StyledDiv then styled-components gives warning that there are frequently changed styles for same component and advise to use second version.

Difference in TBT is extreme and function rehydrateSheet takes with first version of StyledDiv and 6x slow CPU **5s** to execute if used with second component it is about **9ms**. It means that if styles are frequently changed it has huge effect into performance until used in second component way, is this true? 
Our codebase has frequently changed styles for same components (not in such extreme way) but I will investigate more. Hope this helps. (Maybe I did not notice something in documentation :) sorry if true)

Thanks Jakub

**Comment by Ghost-str** [comment]
Hi, we have same problem. InsertBefore takes 1 second total when hydrated. 
![Снимок экрана от 2020-10-12 10-24-25](https://user-images.githubusercontent.com/24704554/95717237-50b9e880-0c75-11eb-9aa3-5a9b42563259.png)

**Comment by kitten** [comment]
You're using and benchmarking the non-speedy insertion methods which aren't designed to be particularly fast.

If you have `process.env.NODE_ENV` compiled to `"production"` in, well, production it switches to speedy injection with CSSOM instead and all these numbers will be different.

So overall we can't do anything about this, because it's basically a report of it being slow in development, which is expected.
We may remove the non-speedy injection in the future however so watch out for that.

**Comment by RDemian** [comment]
the problem is not the speed of injection. All styles are already coming from the server side. On the client side, styles are parsed, compared with those that came from the server, and completely overwrite the styles from the server. For some reason, they completely overwrite the styles from the server?

**Comment by kitten** [comment]
@RDemian thats expected behaviour. It's our rehydration process.

**Comment by Sawan1994** [comment]
@kitten We have been using styled-components in our production build for quite some time. Recently, we started looking into performance improvement and styled-components v4 seemed to be one of the main culprits in impacting performance. We moved our web app to styled-components v5 hoping it will improve the performance. There was some performance improvement when we migrated to v5 but after a month we started seeing UI issues as we added new components and pages. So, we again had to move back to styled components v4. 

We have done a POC as well on one of our pages comparing tailwindCSS with styled-components and it drastically improved page performance. TailwindCSS is atleast 25-30 points faster than styled components v4 and 15 - 20 points faster than v5. Which trust me is a big deal as Google's new ranking algorithm is about to introduce performance of page as one of the main metrics in page ranking.
